lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/DFSEvaluatorPreserver.java

```diff
@@ -0,0 +1,183 @@
+
+    final String newContent = outputStream.toString() + appendEntry;
+    fs.delete(path, true);
+
+    try (final FSDataOutputStream newOutput = fs.create(path);
+         final InputStream newInput = new ByteArrayInputStream(newContent.getBytes())) {
+      IOUtils.copyBytes(newInput, newOutput, 4096, true);
+    }
+
+  }
+
+  private String getChangeLogLocation() {
+    return "/ReefApplications/" + EvaluatorManager.getJobIdentifier() + "/evaluatorsChangesLog";
+  }
+
+  private void logContainerAddition(final String containerId) {
+    final String entry = ADD_FLAG + containerId + System.lineSeparator();
+    logContainerChange(entry);
+  }
+
+  private void logContainerRemoval(final String containerId) {
+    final String entry = REMOVE_FLAG + containerId + System.lineSeparator();
+    logContainerChange(entry);
+  }
+
+  private void logContainerChange(final String entry) {
+    try {
+      writeToEvaluatorLog(entry);
+    } catch (final IOException e) {
+      final String errorMsg = "Unable to log the change of container [" + entry +
+          "] to the container log. Driver restart won't work properly.";
+      LOG.log(Level.WARNING, errorMsg, e);
+      throw new RuntimeException(errorMsg);
+    }
+  }
+
+  private Set<String> getExpectedContainersFromLogReplay() {
+    final org.apache.hadoop.conf.Configuration config = new org.apache.hadoop.conf.Configuration();
+    config.setBoolean("dfs.support.append", true);
+    config.setBoolean("dfs.support.broken.append", true);
+    final Set<String> expectedContainers = new HashSet<>();
+    try {
+      final FileSystem fs = FileSystem.get(config);
+      final Path path = new Path(getChangeLogLocation());
+      if (!fs.exists(path)) {
+        // empty set
+        return expectedContainers;
+      } else {
+        final BufferedReader br = new BufferedReader(new InputStreamReader(fs.open(path)));
+        String line = br.readLine();
+        while (line != null) {
+          if (line.startsWith(ADD_FLAG)) {
+            final String containerId = line.substring(ADD_FLAG.length());
+            if (expectedContainers.contains(containerId)) {
+              throw new RuntimeException("Duplicated add container record found in the change log for container " +
+                  containerId);
+            }
+            expectedContainers.add(containerId);
+          } else if (line.startsWith(REMOVE_FLAG)) {
+            final String containerId = line.substring(REMOVE_FLAG.length());
+            if (!expectedContainers.contains(containerId)) {
+              throw new RuntimeException("Change log includes record that try to remove non-exist or duplicate " +
+                  "remove record for container + " + containerId);
+            }
+            expectedContainers.remove(containerId);
+          }
+          line = br.readLine();
+        }
+        br.close();
+      }
+    } catch (final IOException e) {
+      throw new RuntimeException("Cannot read from log file", e);
```

Comment:
Is there another reaction possible? Some apps might be able to recover in a restart even if the evaluator log can't be read?